File: exam/02/home/hachiroku_go_1.cc
// alternative provided by TA 李旺陽
void dfs(long v, const vector<vector<long>> &E, vector<bool> &used) {
    used[v] = true;
    for(long e:E[v])
        if(!used[e])
            dfs(e, E, used);
}

string Travle1(long n, long m, vector<vector<long>> edges) {
    vector<vector<long>> E(n);
    vector<long> deg(n);
    vector<bool> used(n);

    for(auto &e:edges) {
        E[e[0]].emplace_back(e[1]);
        E[e[1]].emplace_back(e[0]);
        deg[e[0]]++;
        deg[e[1]]++;
    }

    dfs(0, E, used);

    if ( any_of(deg.begin(), deg.end(), [](long i){ return i % 2 == 1; }) ||
         count(used.begin(), used.end(), false) > 0 )
        return "no";
    return "yes";
}

// alternative provided by TA 鄧晉杰
template <typename Graph>
long BFS (Graph const &graph)
{
  long count {};
  std::vector<bool> is_discovered(graph.size());
  std::deque<long> nodes;
  is_discovered[0] = true;
  nodes.emplace_back(0);
  while (!nodes.empty())
  {
    ++count;
    auto from {nodes.front()}; nodes.pop_front();
    for (auto const to : graph[from])
    {
      if (!is_discovered[to])
      {
        is_discovered[to] = true;
        nodes.emplace_back(to);
      }
    }
  }
  return count;
}

std::string Travle1 (long n, long m, std::vector<std::vector<long>> edges)
{
  std::vector<std::vector<long>> graph(n);
  for (auto const &edge : edges)
  {
    graph[edge[0]].emplace_back(edge[1]);
    graph[edge[1]].emplace_back(edge[0]);
  }
  for (auto neighbors : graph)
  {
    if ((neighbors.size() == 0) || (neighbors.size() % 2 == 1))
    {
      return "no";
    }
  }
  if (BFS(graph) != n)
  {
    return "no";
  }
  return "yes";
}

File: exam/02/home/hachiroku_go_1.py
# alternative provided by TA 李旺陽
def Travle1(n, m, edges):
    deg = [ 0 for _ in range(n) ]
    E = [ [] for _ in range(n) ]
    for a, b in edges:
        E[a].append(b)
        E[b].append(a)
        deg[a] = 1 - deg[a]
        deg[b] = 1 - deg[b]

    if 1 in deg:
        return "no"

    used = [ False for _ in range(n) ]
    st = [0]
    used[0] = True

    while len(st) > 0:
        v = st.pop()
        for e in E[v]:
            if not used[e]:
                used[e] = True
                st.append(e)
    return "yes" if False not in used else "no"
